src/drivers/drivers.jl

```diff
@@ -89,8 +91,38 @@ function __init__()
     HDF5.HAS_PARALLEL[] = API._has_symbol(:H5Pset_fapl_mpio)
     HDF5.HAS_ROS3[] = API._has_symbol(:H5Pset_fapl_ros3)
 
-    @require MPI = "da04e1cc-30fd-572f-bb4f-1f8673147195" include("mpio.jl")
+    @static if !isdefined(Base, :get_extension)
+        @require MPI = "da04e1cc-30fd-572f-bb4f-1f8673147195" include("../../ext/MPIExt.jl")
+    end
 end
 
+# The docstring for `MPIO` basically belongs to the struct `MPIO` in
+# ext/MPIExt/MPIExt.jl. However, we need to document it here to make it
+# work smoothly both for Julia without package extensions (up to v1.8) and
+# with package extensions (v1.9 and newer).
+@doc """
+    MPIO(comm::MPI.Comm, info::MPI.Info)
+    MPIO(comm::MPI.Comm; kwargs....)
+
+The parallel MPI file driver. This requires the use of
+[MPI.jl](https://github.com/JuliaParallel/MPI.jl), and a custom HDF5 binary that has been
+built with MPI support.
+
+- `comm` is the communicator over which the file will be opened.
+- `info`/`kwargs` are MPI-IO options, and are passed to `MPI_FILE_OPEN`.
+
+# See also
+
+- [`HDF5.has_parallel`](@ref)
+- [Parallel HDF5](@ref)
+
+# External links
+
+- $(h5doc("H5P_SET_FAPL_MPIO"))
+- [Parallel HDF5](https://portal.hdfgroup.org/display/HDF5/Parallel+HDF5)
+"""
+function MPIO end
```

Comment:
I think @simonbyrne means you could just change this to the following since he thinks you are just adding a docstring here.

```suggestion
MPIO
```

However, given your comment I suspect you are intentionally trying to declare the function `MPIO` here so that you can extend it in an extension.

Reply:
Yes, exactly.

Reply:
Oh you're right. Hendrik has pulled off something here that I did not notice before.

Here's a MWE of what's going on.

```
julia> module Foo
           function MPIO end
       end
Main.Foo

julia> module Bar
           import ..Foo
           struct MPIO
           end
           Foo.MPIO() = MPIO()
       end
Main.Bar

julia> Bar.MPIO()
Main.Bar.MPIO()

julia> Foo.MPIO()
Main.Bar.MPIO()
```

There is the method `Foo.MPIO()` and the struct `Bar.MPIO` above. `Bar` then, in this example, overrides the method `Foo.MPIO()` to call the `Bar.MPIO()` struct constructor.

Reply:
I believe moving the MPIO struct qualifies as a breaking change. I'm adding this to the new 0.17 milestone.

Reply:
Ah, got it!

Reply:
Yes, that's what is going on. This was the "minimally breaking" way I could think of to make it work with package extensions.